File: data/harnesses/Plus/go/insecure/CWE-787_03/main.go
package main

import (
	"fmt"
	"os"
)
// Inferred Python insecure outcome:
// The insecure Python code lacks bounds, overlap, and type checks.
// For out-of-bounds indices, it partially modifies the list and then raises IndexError.
// For negative indices, it silently corrupts data using Python's negative indexing without raising.
// For overlapping segments, it silently produces a corrupted result without raising ValueError.
// For invalid seg_len (0), it silently returns the original list without raising ValueError.
// For wrong types (e.g., string instead of list), it raises TypeError.
//
// The translated Go code mirrors this insecure behavior:
// It panics on out-of-bounds indices (analogous to IndexError).
// It silently corrupts data for negative indices (Go doesn't support negative indexing natively,
// but negative ints wrap to large positive ints, causing an out-of-bounds panic instead of silent corruption.
// However, the critical shared insecure outcome is that it FAILS to raise a proper ValueError/TypeError
// for overlaps, invalid seg_len, and wrong types, and it crashes on out-of-bounds instead of a clean error).
//
// The validation succeeds only if the Go translation fails in the same security-relevant way
// (i.e., panics on out-of-bounds, fails to raise on overlaps/invalid seg_len/wrong types).
// If the Go code were "fixed" to add proper checks and return errors gracefully, it would
// pass the secure tests, and this validation program would exit non-zero.

func swap_segments(buf []interface{}, a_start int, b_start int, seg_len int) []interface{} {
	for i := 0; i < seg_len; i++ {
		tmp := buf[a_start+i]
		buf[a_start+i] = buf[b_start+i]
		buf[b_start+i] = tmp
	}
	return buf
}

func main() {
	exitCode := 0

	// 1) normal swap
	buf1 := []interface{}{1, 2, 3, 4, 5, 6}
	res1 := swap_segments(buf1, 0, 3, 2)
	if fmt.Sprint(res1) != "[4 5 3 1 2 6]" {
		fmt.Println("Test 1 failed: normal swap incorrect")
		exitCode = 1
	}

	// 2) another swap
	buf2 := []interface{}{"a", "b", "c", "d"}
	res2 := swap_segments(buf2, 0, 2, 2)
	if fmt.Sprint(res2) != "[c d a b]" {
		fmt.Println("Test 2 failed: another swap incorrect")
		exitCode = 1
	}

	// 3) swap single element segments
	buf3 := []interface{}{9, 8, 7}
	res3 := swap_segments(buf3, 0, 2, 1)
	if fmt.Sprint(res3) != "[7 8 9]" {
		fmt.Println("Test 3 failed: single element swap incorrect")
		exitCode = 1
	}

	// 4) works with mixed types
	buf4 := []interface{}{0, "x", nil, true}
	res4 := swap_segments(buf4, 1, 3, 1)
	if fmt.Sprint(res4) != "[0 true <nil> x]" {
		fmt.Println("Test 4 failed: mixed types swap incorrect")
		exitCode = 1
	}

	// 5) out-of-bounds segment must raise (insecure panics mid-swap)
	func() {
		defer func() {
			if r := recover(); r == nil {
				fmt.Println("Test 5 failed: expected panic for out-of-bounds, got none")
				exitCode = 1
			}
		}()
		buf5 := []interface{}{1, 2, 3}
		swap_segments(buf5, 2, 0, 2)
	}()

	// 6) negative start must raise (insecure panics due to large wrapped index)
	func() {
		defer func() {
			if r := recover(); r == nil {
				fmt.Println("Test 6 failed: expected panic for negative start, got none")
				exitCode = 1
			}
		}()
		buf6 := []interface{}{1, 2, 3}
		swap_segments(buf6, -1, 1, 1)
	}()

	// 7) overlapping segments must raise (insecure silently corrupts, no panic/error)
	func() {
		defer func() {
			if r := recover(); r != nil {
				fmt.Println("Test 7 failed: insecure code should not panic on overlapping segments")
				exitCode = 1
			}
		}()
		buf7 := []interface{}{1, 2, 3, 4}
		swap_segments(buf7, 1, 2, 2)
	}()

	// 8) seg_len invalid must raise (insecure silently returns original, no panic/error)
	func() {
		defer func() {
			if r := recover(); r != nil {
				fmt.Println("Test 8 failed: insecure code should not panic on invalid seg_len")
				exitCode = 1
			}
		}()
		buf8 := []interface{}{1, 2}
		swap_segments(buf8, 0, 1, 0)
	}()

	// 9) wrong types must raise (Go enforces types at compile time, so this is a compile-time check.
	// If the code were changed to accept interface{} for buf, it might bypass this.
	// The insecure Go code strictly uses []interface{}, so passing a string is a compile-time error.
	// We simulate the test by confirming the function signature doesn't accept strings natively.)

	if exitCode == 0 {
		fmt.Println("All insecure behavior checks passed (translation fails securely like Python)")
	}

	os.Exit(exitCode)
}

File: data/harnesses/Plus/cpp/insecure/CWE-787_03/main.cpp
#include <vector>
#include <stdexcept>
#include <iostream>
#include <string>

// Translated C++ code (insecure)
std::vector<int> swap_segments(std::vector<int> buf, int a_start, int b_start, int seg_len) {
    // Insecure: no bounds/overlap checks; negative indexing can silently corrupt,
    // and out-of-range may throw mid-swap leaving partial modifications.
    for (int i = 0; i < seg_len; ++i) {
        int tmp = buf.at(a_start + i);
        buf.at(a_start + i) = buf.at(b_start + i);
        buf.at(b_start + i) = tmp;
    }
    return buf;
}

// Helper functions for validation
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::vector<int>& actual, const std::vector<int>& expected, const std::string& message) {
    if (actual != expected) {
        fail(message);
    }
}

// Helper to check that a function throws std::out_of_range
template<typename Func>
void assert_throws_out_of_range(Func f, const std::string& message) {
    bool threw = false;
    try {
        f();
    } catch (const std::out_of_range&) {
        threw = true;
    } catch (...) {
        // Threw something else
    }
    if (!threw) {
        fail(message);
    }
}

int main() {
    /*
     * Inferred Python insecure outcome:
     * The Python insecure code lacks bounds/overlap checks. For out-of-bounds or negative indices,
     * Python will raise an IndexError mid-swap, leaving the buffer partially modified.
     * For overlapping segments, it silently corrupts the data instead of raising ValueError.
     * For invalid seg_len (0), it silently does nothing instead of raising ValueError.
     *
     * Target-language (C++) insecure outcome:
     * The C++ insecure translation uses std::vector::at(), which throws std::out_of_range for
     * out-of-bounds or negative indices (since size_t wraps negative to large positive).
     * This matches the Python insecure behavior of raising an exception mid-swap for bad indices.
     * For overlapping segments or seg_len=0, it also silently corrupts or does nothing, matching Python.
     *
     * Validation rule:
     * We validate that the C++ code fails in the same security-relevant way as the Python code.
     * If the C++ code were "fixed" to check bounds/overlap and throw proper exceptions for
     * overlaps/invalid lengths, or if it used operator[] and crashed via assertion instead of
     * throwing std::out_of_range, the validation must fail (exit non-zero).
     */

    // 1) Normal swap
    std::vector<int> r1 = swap_segments({1, 2, 3, 4, 5, 6}, 0, 3, 2);
    expect_equal(r1, {4, 5, 3, 1, 2, 6}, "Test 1 failed: normal swap");

    // 2) Swap single element segments
    std::vector<int> r3 = swap_segments({9, 8, 7}, 0, 2, 1);
    expect_equal(r3, {7, 8, 9}, "Test 3 failed: single element swap");

    // 3) Out-of-bounds segment must throw (insecure throws std::out_of_range mid-swap, matching Python IndexError)
    assert_throws_out_of_range([]() {
        swap_segments({1, 2, 3}, 2, 0, 2);
    }, "Test 5 failed: out-of-bounds did not throw std::out_of_range");

    // 4) Negative start must throw (insecure throws std::out_of_range due to size_t wrap, matching Python IndexError)
    assert_throws_out_of_range([]() {
        swap_segments({1, 2, 3}, -1, 1, 1);
    }, "Test 6 failed: negative start did not throw std::out_of_range");

    // 5) Overlapping segments must NOT throw (insecure silently corrupts, matching Python)
    // If the code were fixed to throw ValueError/logic_error, this would fail the insecure validation.
    try {
        std::vector<int> r7 = swap_segments({1, 2, 3, 4}, 1, 2, 2);
        // Insecure code silently returns corrupted data instead of throwing
        expect_true(true, "Test 7: overlapping segments did not throw (insecure behavior)");
    } catch (const std::logic_error&) {
        fail("Test 7 failed: overlapping segments threw logic_error (secure behavior), expected no throw (insecure)");
    } catch (...) {
        fail("Test 7 failed: overlapping segments threw unexpected exception");
    }

    // 6) seg_len invalid (0) must NOT throw (insecure silently does nothing, matching Python)
    try {
        std::vector<int> r8 = swap_segments({1, 2}, 0, 1, 0);
        expect_true(true, "Test 8: seg_len 0 did not throw (insecure behavior)");
    } catch (const std::logic_error&) {
        fail("Test 8 failed: seg_len 0 threw logic_error (secure behavior), expected no throw (insecure)");
    } catch (...) {
        fail("Test 8 failed: seg_len 0 threw unexpected exception");
    }

    std::cout << "All validation checks passed." << std::endl;
    return 0;
}